File: python/cards.py
import re

def cards_cmp(a, b):
    """
    Sorting helper, returns zero if a == b, a positive integer if
    a > b and a negative integer otherwise.
    """
    val_a, val_b = a[0], b[0]
    for val, denom in zip(['T', 'J', 'Q', 'K', 'A'], range(10, 15)):
        if val_a == val:
            val_a = denom
        if val_b == val:
            val_b = denom
    val_a = int(val_a)
    val_b = int(val_b)
    return val_a - val_b

def denom(card):
    """
    The denominaton of the given card.  The denomination is defined as
    follows:
        2 through 10 - the cards value
        J            - 11
        Q            - 12
        K            - 13
        A            - 14
    """
    val = card[0]
    try:
        return { 'T': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14 }[val]
    except KeyError:
        return int(val)

class Hand(object):
    """
    Stores a poker hand.  Poker hand contains five cards from the deck
    of 52 playing cards.  This class also stores information on
    how many cards of the same denomination are there in the hand.
    """
    def __init__(self, cards):
        """
        Creates new poker hand.  The argument must be a list with cards
        represented as strings.
        """
        cards.sort(cards_cmp)
        self._handlers = [self.royal_flush, self.straight_flush,
                          self.four_of_a_kind, self.full_house, self.flush,
                          self.straight, self.three_of_a_kind,
                          self.two_pairs, self.one_pair]
        self._cards = cards
        self._stretches = []
        self._value = None

    def n_of_a_kind(self, n):
        """
        Returns True if the hand contains exactly n cards of the same
        denomination.
        """
        if not self._stretches:
            counter = 1
            for a, b in zip(self._cards, self._cards[1:] + ['X']):
                if a[0] != b[0]:
                    self._stretches.append(counter)
                    counter = 1
                else: counter += 1
        return n in self._stretches
        
    def royal_flush(self):
        """
        Returns True if all cards in the hand are of the same suit and
        the smallest card is a ten.
        """
        return self._cards[0][0] == 'T' and self.straight_flush()

    def straight_flush(self):
        """
        Similar to royal flush, except cards should be of consequtive 
        denominations (no need to start at 10).
        """
        return self.straight() and self.flush()
    
    def four_of_a_kind(self):
        """
        Retruns True if the hand contains exactly four cards of the same 
        denomination.
        """
        return self.n_of_a_kind(4)
    
    def full_house(self):
        """
        Returns True if the hand contains a pair and a triple of the same 
        denomination.
        """
        if self.one_pair():
            return self._stretches == [2, 3] or self._stretches == [3, 2]
        return False
    
    def flush(self):
        """
        Returns True if all cards in the hand come from the same suit.
        """
        return bool(re.match(r'.(.)(.\1){4}', ''.join(self._cards)))
    
    def straight(self):
        """
        Returns True if cards' values have no ``gaps'' in between.
        """
        prev = denom(self._cards[0])
        for card in self._cards[1:]:
            if not denom(card) == prev + 1:
                return False
            prev += 1
        return True
    
    def three_of_a_kind(self):
        """
        Returns True if the hand has a triple.
        """
        return self.n_of_a_kind(3)
    
    def two_pairs(self):
        """
        Returns True if the hand has two different pairs.
        """
        if self.n_of_a_kind(2):
            return self._stretches.count(2) == 2
        return False
    
    def one_pair(self):
        """
        Returns True if the hand has one pair.
        """
        return self.n_of_a_kind(2)

    def value(self):
        """
        Computes the value of the hand.  Hands are assigned value based on
        some special combinations, such as ``Royal Flush'', ``Three Of A Kind''
        etc, or by taking the denomination of the highest-ranked card.

        See the documentation of other methods of this class for more details
        on the meaning of special combinations.
        """
        if not self._value:
            for v, f in zip(range(24, 10, -1), self._handlers):
                if f():
                    self._value = v
                    break
            if not self._value:
               self._value = denom(self._cards[-1]) 
        return self._value

    def scoring_cards(self, score):
        """
        When two hands have the same value, this procedure is used to compute
        the cards which may potentially affect the tie break.  The cards are
        returned in decreasing order (i.e. the first card should be 
        considered first to break the tie, then the second and so on).
        """
        result = []
        if score in [15, 16, 17, 20, 21]:
            j = 0
            for i in self._stretches:
                if i > 1:
                    result += self._cards[j:j + i]
                j += i
        else:
            result = self._cards[:]
        return reversed(result)

    def break_tie(self, other):
        """
        When two hands have the same value, this is used to break ties.  Note that
        ties can still occur because the suit plays no role in this process.
        """
        val = self.value()
        scoring_a = self.scoring_cards(val)
        scoring_b = other.scoring_cards(val)
        for a, b in zip(scoring_a, scoring_b):
            da, db = denom(a), denom(b)
            if da != db:
                return da - db
        return 0

    def __lt__(self, other):
        """
        Implements the less than operator.
        """
        val_a, val_b = self.value(), other.value()
        if val_a == val_b:
            return self.break_tie(other) < 0
        return val_a < val_b

    def __le__(self, other):
        """
        Implements the less than or equal operator.
        """
        val_a, val_b = self.value(), other.value()
        if val_a == val_b:
            return self.break_tie(other) <= 0
        return val_a <= val_b
    
    def __eq__(self, other):
        """
        Implements the equality operator.
        """
        if self.value() == other.value():
            return self.break_tie(other) == 0
        return False
    
    def __ne__(self, other):
        """
        Implements the not equal operator.
        """
        return not self.__eq__(other)
    
    def __gt__(self, other):
        """
        Implements the greater than operator.
        """
        return not self.__le__(other)
    
    def __ge__(self, other):
        """
        Implements the greater or equal operator.
        """
        return not self.__lt__(other)

File: cxx/cards.cxx
/*! \file combinatoricx.cxx
 *  This file contains some commonly used functions over integers.
 */
#include <vector>
#include <cmath>
#include <algorithm>
#include <string>

/*! \struct card_t
 *  \brief This struct represents a playing card from a 52-cards deck.
 *  \var suit is the card's suit.
 *  \var denom is the card's denomination.
 */
struct card_t {
    unsigned char suit;
    
    unsigned char denom;
    
    card_t() : suit(' '), denom(' ') {}
    
    card_t(unsigned char suit_, unsigned char denom_) : suit(suit_), denom(denom_) {}
};

/*! \fn bool card_cmp(card_t a, card_t b)
 *  \brief This is a helper function used to sort cards based on their denomination.
 *  \param a The card prior in the sequence.
 *  \param b The card following in the sequence.
 *  \returns true if a must precede b.
 *
 *  Helps `std::sort()` to sort a collection of `card_t` based on card's denomination.
 */
bool card_cmp(card_t a, card_t b) { return a.denom < b.denom; }

/*! \fn std::vector<card_t> parse_cards(std::string raw)
 *  \brief Parses hand (5 cards) from string.
 *  \param raw Taw string containing 5 card codes.
 *  \returns A vector with parsed cards.
 *
 *  Given a string with five card codes will create a sorted vector with five
 *  cards.  Cards are sorted by their denomination.
 */
std::vector<card_t> parse_cards(std::string raw) {
    std::vector<card_t> result(5);
    int state = 0;
    unsigned char suit, denom;
    std::string denoms("123456789TJQKA");
    std::size_t place = 0;
    std::string normalized = raw[raw.size() - 1] == ' ' ? raw : raw + " ";
    
    for (auto c : normalized) {
        switch (state) {
            case 0:
                denom = denoms.find(c) + 1;
                state = 1;
                break;
            case 1:
                suit = (unsigned char)c;
                state = 2;
                break;
            case 2:
                struct card_t created(suit, denom);
                result[place++] = created;
                state = 0;
        }
    }
    std::sort(result.begin(), result.end(), card_cmp);
    return result;
}

/*! \fn std::vector<unsigned char> group_cards(std::vector<card_t> cards)
 *  \brief Groups cards by denomination.
 *  \param cards Vector containing a sorted sequence of cards (not modified).
 *  \returns A vector with stretches of cards of the same denomination.
 *
 *  This function implements something very similar to run-length encoding.
 *  Even though this is a canonical group-by functionality, this function
 *  exploits the fact that the input sequence is sorted, so it is trivially
 *  linear.
 */
std::vector<unsigned char> group_cards(std::vector<card_t> cards) {
    std::vector<unsigned char> result;
    unsigned char counter = 1;

    for (std::size_t t = 1; t < cards.size(); t++) {
        if (cards[t - 1].denom == cards[t].denom) {
            counter++;
        } else {
            result.push_back(counter);
            counter = 1;
        }
    }
    result.push_back(counter);
    return result;
}

/*! \class hand
 *  \brief This class provides operations on poker hand.
 *
 *  This class provides methods assessing the value of one's poker hand.
 */
class hand {
    std::vector<card_t> cards;
    
    std::vector<unsigned char> stretch;

  public:
  
    /*! \fn hand(std::string raw)
     *  \brief Creates new poker hand.
     *  \param raw Raw poker hand.
     *
     *  Constructor will parse the hand from the given string and initialize its
     *  fields based on the parsing results.
     */
    hand(std::string raw) :
            cards(parse_cards(raw)),
            stretch(group_cards(cards)) { }

    /*! \fn bool n_of_a_kind(int n) const
     *  \brief Verifies if the hand has `n` cards of the same denomination.
     *  \param n Number of same cards needed to be found.
     *  \returns true iff the hand has this exact number of cards of the same
     *  denomination.
     *
     *  Looks for at least one stretch of the required size.  Note that the hand
     *  that contains 3 cards with the same denomination will not be considered
     *  as having 2 cards of the same denomination, unless it also contains
     *  exactly 2 cards of the same denomination.  Other methods rely on this
     *  behavior.
     */
    bool n_of_a_kind(int n) const {
        return std::find(stretch.begin(), stretch.end(), n) != stretch.end();
    }

    /*! \fn bool royal_flush() const
     *  \brief Verifies whether the hand has a T,J,Q,K,A sequence all of the
     *  same suit.
     *  \returns true iff the hand contains the ``Royal Flush'' combination.
     */
    bool royal_flush() const {
        return cards[0].denom == 10 && straight_flush();
    }

    /*! \fn bool straight_flush() const
     *  \brief Verifies whether the hand has five cards of consequent
     *  denomination, all of same suit.
     *  \returns true iff the hand contains the ``Straight Flush'' combination.
     */
    bool straight_flush() const { return straight() && flush(); }

    /*! \fn bool four_of_a_kind() const
     *  \brief Verifies whether the hand contains four card of the same
     *  denomination.
     *  \returns true iff the hand contains the ``Four Of A Kind'' combination.
     */
    bool four_of_a_kind() const { return n_of_a_kind(4); }

    /*! \fn bool full_house() const
     *  \brief Verifies whether the hand contains a pair and that other three
     *  cards are of the same denomination.
     *  \returns true iff the hand contains the ``Full House'' combination.
     */
    bool full_house() const {
        if (one_pair()) {
            return (stretch[0] == 2, stretch[1] == 3 ||
                    stretch[1] == 2, stretch[0] == 3);
        }
        return false;
    }

    /*! \fn bool flush() const
     *  \brief Verifies whether all cards in the hand are of the same suit.
     *  \returns true iff the hand contains the ``Flush'' combination.
     */
    bool flush() const {
        unsigned char elt = cards[0].suit;
        return std::count_if(
            cards.begin(), cards.end(),
            [elt](card_t x) -> bool { return x.suit == elt; }) == cards.size();
    }

    /*! \fn bool straight() const
     *  \brief Verifies that there are no two cards of the same denomination and
     *  that the difference between the highest-valued card and the lowest-valued
     *  one is exactly five.
     *  \returns true iff the hand contains the ``straight'' combination.
     */
    bool straight() const {
        for (std::size_t t = 1; t < cards.size(); t++) {
            if (cards[t].denom != cards[t - 1].denom + 1) {
                return false;
            }
        }
        return true;
    }

    /*! \fn bool three_of_a_kind() const
     *  \brief Verifies whether the hand contains exactly three cards of the
     *  same denomination.
     *  \returns true iff the hand contains the ``Three Of A Kind'' combination.
     */
    bool three_of_a_kind() const { return n_of_a_kind(3); }

    /*! \fn bool two_pairs() const
     *  \brief Verifies whether the hand contains two different pairs of cards
     *  of the same denomination.
     *  \returns true iff the hand contains the ``Two Pairs'' combination.
     */
    bool two_pairs() const {
        if (n_of_a_kind(2)) {
            return std::count(stretch.begin(), stretch.end(), 2) == 2;
        }
        return false;
    }

    /*! \fn bool one_pair() const
     *  \brief Verifies whether the hand contains a pair of cards
     *  of the same denomination.
     *  \returns true iff the hand contains the ``One Pair'' combination.
     */
    bool one_pair() const { return n_of_a_kind(2); }

    /*! \fn int value() const
     *  \brief Calculates the value of the hand.
     *  \returns An integer in range [3,23] indicating the total value of the hand.
     *
     *  The value of the hand is determined by it either having one of the eight
     *  known special combinations, or by its highest-ranked card.  The combinations
     *  are ranked as follows:
     *  - *One Pair*:        15
     *  - *Two Pairs*:       16
     *  - *Three Of A Kind*: 17
     *  - *Straight*:        18
     *  - *Flush*:           19
     *  - *Full House*:      20
     *  - *Four Of A Kind*:  21
     *  - *Straight Flush*:  22
     *  - *Royal Flush*:     23
     *  .
     */
    int value() const {
        bool result;
        for (std::size_t t = 0; t < 9; t++) {
            switch (t) {
                case 0: result = royal_flush();     break;
                case 1: result = straight_flush();  break;
                case 2: result = four_of_a_kind();  break;
                case 3: result = full_house();      break;
                case 4: result = flush();           break;
                case 5: result = straight();        break;
                case 6: result = three_of_a_kind(); break;
                case 7: result = two_pairs();       break;
                case 8: result = one_pair();        break;
            }
            if (result) {
                return 15 + (8 - t);
            }
        }
        return (int)cards[cards.size() - 1].denom;
    }

    /*! \fn std::vector<card_t> scoring_cards() const
     *  \brief Returns only the cards that can affect a tie-break.
     *  \returns A vector of cards which can affect a tie-break.
     *
     *  Ties are broken by first considering the cards that participate in
     *  forming one of the special combinations \see hand::value.  The cards
     *  are also considered in the order from highest to lowest value.
     */
    std::vector<card_t> scoring_cards() const {
        std::vector<card_t> result;
        
        switch (value()) {
            case 15: case 16: case 17: case 20: case 21:
                {
                    std::size_t j = 0;
                    for (int s : stretch) {
                        if (s > 1) {
                            result.resize(result.size() + s);
                            std::copy(&cards[j], &cards[j + s],
                                      &result[result.size() - s]);
                        }
                        j += s;
                    }
                }
                break;
            default:
                result.reserve(cards.size());
                std::copy(cards.begin(), cards.end(), result.begin());
        }
        std::reverse(result.begin(), result.end());
        return result;
    }

    /*! \fn int break_tie(const hand& other) const
     *  \brief Computes amount of points by which one hand outweighs the other.
     *  \returns The number of points by which this hand outweighs the other.
     *
     *  Once the tie is detected, this function is given the `other` hand, which
     *  has the same score, but might have different cards supporting that score.
     *  Ties are broken by first looking into the highest-ranked cards in the
     *  special combinations and later by looking at the remaining cards.
     *
     *  Note that it is still possible to draw (we don't establish any rules which
     *  assign weights to suits).
     */
    int break_tie(const hand& other) const {
        std::vector<card_t> scoring_a = scoring_cards();
        std::vector<card_t> scoring_b = other.scoring_cards();

        for (std::size_t t = 0; t < scoring_a.size(); t++) {
            if (scoring_a[t].denom != scoring_b[t].denom) {
                return scoring_a[t].denom - scoring_b[t].denom;
            }
        }
        return 0;
    }

    /*! \fn bool operator >(const hand& other) const
     *  \brief Establishes whether this hand's value is greater than the
     *  value of the other hand.
     *  \returns `true` iff the value of this hand is greate than that of the other.
     *
     *  Invoking this operator will set in motion `hand::value()` and possibly
     *  `hand::break_tie()` in order to establish the winning hand.
     */
    bool operator >(const hand& other) const {
        int a = value();
        int b = other.value();
        if (a == b) a += break_tie(other);
        return a > b;
    }
};
